lib/aliquot/payment.rb

```diff
@@ -0,0 +1,134 @@
+      Aliquot.start_key_updater(logger) if $key_updater_thread.nil? && signing_keys.nil?
+
+      validation = Aliquot::Validator::Token.new(JSON.parse(token_string))
+      validation.validate
+
+      @token = validation.output
+
+      @shared_secret = shared_secret
+      @merchant_id   = merchant_id
+      @signing_keys  = signing_keys
+    end
+
+    ##
+    # Validate and decrypt the token.
+    def process
+      unless valid_protocol_version?
+        raise Error, 'only ECv1 protocolVersion is supported'
+      end
+
+      unless valid_signature?
+        raise InvalidSignatureError
+      end
+
+      validator = Aliquot::Validator::SignedMessage.new(JSON.parse(@token['signedMessage']))
+      validator.validate
+      signed_message = validator.output
+
+      aes_key, mac_key = derive_keys(signed_message['ephemeralPublicKey'], @shared_secret, 'Google')
+
+      unless self.class.valid_mac?(mac_key, signed_message['encryptedMessage'], signed_message['tag'])
+        raise InvalidMacError
+      end
+
+      @message = JSON.parse(decrypt(aes_key, signed_message['encryptedMessage']))
+
+      Aliquot::Validator::EncryptedMessageValidator.new(@message).validate
+
+      raise ExpiredException if expired?
+
+      @message
+    end
+
+    def protocol_version
+      @token['protocolVersion']
+    end
+
+    def valid_protocol_version?
+      protocol_version == 'ECv1'
+    end
+
+    ##
+    # Check if the token is expired, according to the messageExpiration included
+    # in the token.
+    def expired?
+      @message['messageExpiration'].to_f / 1000.0 <= Time.now.to_f
+    end
+
+    def valid_signature?
+      signed_string = ['Google', @merchant_id, protocol_version, @token['signedMessage']].map do |str|
+        [str.length].pack('V') + str
+      end.join
+
+      keys = JSON.parse(signing_keys)['keys']
+      # Check if signature was performed with any possible key.
+      keys.map do |e|
+        next if e['protocolVersion'] != protocol_version
+
+        ec = OpenSSL::PKey::EC.new(Base64.strict_decode64(e['keyValue']))
+        d  = OpenSSL::Digest::SHA256.new
+        ec.verify(d, Base64.strict_decode64(@token['signature']), signed_string)
+      end.any?
+    end
+
+    private
+
+    def derive_keys(ephemeral_public_key, shared_secret, info)
+      ikm = Base64.strict_decode64(ephemeral_public_key) +
+            Base64.strict_decode64(shared_secret)
+      hbytes = HKDF.new(ikm, algorithm: 'SHA256', info: info).next_bytes(32)
+
+      [hbytes[0..15], hbytes[16..32]]
+    end
+
+    def decrypt(key, encrypted)
+      c = OpenSSL::Cipher::AES128.new(:CTR)
+      c.key = key
+      c.decrypt
+
+      c.update(Base64.strict_decode64(encrypted)) + c.final
+    end
+
+    def self.valid_mac?(mac_key, data, tag)
+      digest = OpenSSL::Digest::SHA256.new
+      mac = OpenSSL::HMAC.digest(digest, mac_key, Base64.strict_decode64(data))
+
+      compare(Base64.strict_encode64(mac), tag)
```

Comment:
I am not sure if we should instead Base64 decode `tag` — does it make any difference wrt. the lengths and thus the "constant-time-ness"?

Reply:
It doesn't matter with regards to length, `mac` is always the same length.
Either is fine.

Reply:
But `tag` is taken from the signed message, so that could have incorrect length, no?

Reply:
Hmm. I thought the length was checked in validator, but I see it is not.

However, the tag is of fixed length, 32 bytes and thus 44 base64 characters.

Reply:
Uhm, perhaps the length should be checked by the validator? Nah, then it wouldn't support e.g. `RSAv1` ... Rather we could check the length here.

Reply:
I added a check for it in the validator. :/
Based on the fact that the rest of the validation is very specific to the `ECv1` method.